apps/passport-client/src/dispatch.ts

```diff
@@ -791,25 +794,44 @@ async function loadAfterLogin(
   }
 
   // Validate stored state against the user response.
-  const identityPCD = findUserIdentityPCD(pcds, userResponse.value);
+  const identityPCDV3 = findUserIdentityV3PCD(pcds, userResponse.value);
   if (
     !validateAndLogRunningAppState(
       "loadAfterLogin",
       userResponse.value,
-      identityPCD?.claim?.identity,
+      identityPCDV3?.claim?.identity,
       pcds
     )
   ) {
     userInvalid(update);
     return;
   }
-  if (!identityPCD) {
+  if (!identityPCDV3) {
     // Validation should've caught this, but the compiler doesn't know that.
     console.error("No identity PCD found in encrypted storage.");
     userInvalid(update);
     return;
   }
 
+  const identityV4PCD = findUserIdentityV4PCD(pcds, userResponse.value);
+  if (!identityV4PCD) {
+    try {
```

Comment:
I feel like this migration should ideally run before validateAndLogRunningAppState above so it can validate the migrated state.  It's probably valid to do this earlier.  Or else to call validateAndLogRunningAppState again.
However I think doing the (local) identity migration inside of SemaphoreIdentityPCD deserialization will safely remove some of this complexity.
This code block would still need to exist to call requestAddSemaphoreV4Commitment, but it would be triggered based on the contents of `self` rather than the contents of the PCDCollection.

Reply:
If you don't like the single-PCD approach I'm not going to push it (though I still like it personally).  My alternative suggestion would be to consider how the various different migration points (like this one) can be unified and combined so there isn't migration code spread around various places.  I worry about it making the code complex over time, and of missing bugs if the migration code in different places is subtly different.
One option might be to centralize it all in a single function like ensureSemaphoreV4Migration which can be called in each of the places where v3tov4IdentityPCD and/or requestUpgradeUserWithV4Commitment are called now.  That function might need multiple versions, or multiple subcases internally based on arguments, but reducing the migration down to a single function call in places like this will reduce the overhead in complexity going forward.